File: src/lib/storage.ts
/**
 * Token storage utilities
 * Stores OAuth credentials securely in user's home directory
 */

import os from 'os'
import path from 'path'

export interface ClaudeOAuthCredentials {
  accessToken: string
  refreshToken: string
  expiresAt: number
  connectedAt: number
}

const CONFIG_DIR = path.join(os.homedir(), '.claude-oauth')
const CREDENTIALS_FILE = path.join(CONFIG_DIR, 'credentials.json')

/**
 * Ensure config directory exists
 */
async function ensureConfigDir(): Promise<void> {
  const dir = Bun.file(CONFIG_DIR)
  if (!(await dir.exists())) {
    await Bun.write(path.join(CONFIG_DIR, '.keep'), '')
  }
}

/**
 * Save OAuth credentials to file
 */
export async function saveCredentials(credentials: ClaudeOAuthCredentials): Promise<void> {
  await ensureConfigDir()
  await Bun.write(CREDENTIALS_FILE, JSON.stringify(credentials, null, 2))
  // Set restrictive permissions (owner read/write only)
  if (process.platform !== 'win32') {
    const { chmod } = await import('fs/promises')
    await chmod(CREDENTIALS_FILE, 0o600)
  }
}

/**
 * Load OAuth credentials from file
 * @returns Credentials or null if not found
 */
export async function loadCredentials(): Promise<ClaudeOAuthCredentials | null> {
  try {
    const file = Bun.file(CREDENTIALS_FILE)
    if (!(await file.exists())) {
      return null
    }
    const content = await file.text()
    return JSON.parse(content) as ClaudeOAuthCredentials
  } catch {
    return null
  }
}

/**
 * Clear stored credentials
 */
export async function clearCredentials(): Promise<void> {
  try {
    const { unlink } = await import('fs/promises')
    await unlink(CREDENTIALS_FILE)
  } catch {
    // Ignore if file doesn't exist
  }
}

/**
 * Check if credentials exist and are valid (not expired)
 */
export async function hasValidCredentials(): Promise<boolean> {
  const credentials = await loadCredentials()
  if (!credentials) return false
  
  // Check if access token is expired (with 5 minute buffer)
  const bufferMs = 5 * 60 * 1000
  return credentials.expiresAt > Date.now() + bufferMs
}

/**
 * Get config directory path
 */
export function getConfigDir(): string {
  return CONFIG_DIR
}

File: dotnet-claude-oauth/ClaudeOAuth/Storage.cs
using System.Text.Json;

namespace ClaudeOAuth;

/// <summary>
/// Token storage utilities - stores OAuth credentials securely in user's home directory.
/// </summary>
public class CredentialStorage : ICredentialStorage
{
    private static readonly string ConfigDir = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
        ".claude-oauth"
    );

    private static readonly string CredentialsFile = Path.Combine(ConfigDir, "credentials.json");

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    };

    /// <summary>
    /// Get the config directory path.
    /// </summary>
    public static string GetConfigDir() => ConfigDir;

    /// <summary>
    /// Ensure config directory exists.
    /// </summary>
    private static void EnsureConfigDir()
    {
        if (!Directory.Exists(ConfigDir))
        {
            Directory.CreateDirectory(ConfigDir);
        }
    }

    /// <summary>
    /// Save OAuth credentials to file.
    /// </summary>
    public async Task SaveCredentialsAsync(ClaudeOAuthCredentials credentials)
    {
        EnsureConfigDir();
        var json = JsonSerializer.Serialize(credentials, JsonOptions);
        await File.WriteAllTextAsync(CredentialsFile, json);

        // Set restrictive permissions on non-Windows platforms
        if (!OperatingSystem.IsWindows())
        {
            File.SetUnixFileMode(CredentialsFile, UnixFileMode.UserRead | UnixFileMode.UserWrite);
        }
    }

    /// <summary>
    /// Load OAuth credentials from file.
    /// </summary>
    /// <returns>Credentials or null if not found</returns>
    public async Task<ClaudeOAuthCredentials?> LoadCredentialsAsync()
    {
        try
        {
            if (!File.Exists(CredentialsFile))
            {
                return null;
            }

            var json = await File.ReadAllTextAsync(CredentialsFile);
            return JsonSerializer.Deserialize<ClaudeOAuthCredentials>(json, JsonOptions);
        }
        catch
        {
            return null;
        }
    }

    /// <summary>
    /// Clear stored credentials.
    /// </summary>
    public Task ClearCredentialsAsync()
    {
        try
        {
            if (File.Exists(CredentialsFile))
            {
                File.Delete(CredentialsFile);
            }
        }
        catch
        {
            // Ignore if file doesn't exist
        }

        return Task.CompletedTask;
    }

    /// <summary>
    /// Check if credentials exist and are valid (not expired).
    /// </summary>
    public async Task<bool> HasValidCredentialsAsync()
    {
        var credentials = await LoadCredentialsAsync();
        if (credentials == null) return false;

        // Check if access token is expired (with 5 minute buffer)
        var bufferMs = 5 * 60 * 1000;
        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        return credentials.ExpiresAt > now + bufferMs;
    }
}

/// <summary>
/// Interface for credential storage operations.
/// </summary>
public interface ICredentialStorage
{
    /// <summary>
    /// Save OAuth credentials.
    /// </summary>
    Task SaveCredentialsAsync(ClaudeOAuthCredentials credentials);

    /// <summary>
    /// Load OAuth credentials.
    /// </summary>
    Task<ClaudeOAuthCredentials?> LoadCredentialsAsync();

    /// <summary>
    /// Clear stored credentials.
    /// </summary>
    Task ClearCredentialsAsync();

    /// <summary>
    /// Check if credentials exist and are valid.
    /// </summary>
    Task<bool> HasValidCredentialsAsync();
}
